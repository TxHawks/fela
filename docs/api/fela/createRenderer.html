
<!DOCTYPE HTML>
<html lang="en" >
    <head>
        <title>6.1.1. createRenderer Â· Fela</title>
        <meta charset="UTF-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge" />
        <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
        <meta name="description" content="">
        <meta name="generator" content="GitBook 3.1.0">
        <meta name="author" content="Robin Frischmann <robin@rofrischmann.de>">
        
        
    
    <link rel="stylesheet" href="../../../gitbook/style.css">

    
            
                
                <link rel="stylesheet" href="../../../gitbook/gitbook-plugin-prism/prism.css">
                
            
                
                <link rel="stylesheet" href="../../../gitbook/gitbook-plugin-advanced-emoji/emoji-website.css">
                
            
                
                <link rel="stylesheet" href="../../../gitbook/gitbook-plugin-search/search.css">
                
            
                
                <link rel="stylesheet" href="../../../gitbook/gitbook-plugin-fontsettings/website.css">
                
            
        

    

    
        
    
        
    
        
    
        
    
        
    
        
    

        
    
    
    <meta name="HandheldFriendly" content="true"/>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black">
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../../../gitbook/images/apple-touch-icon-precomposed-152.png">
    <link rel="shortcut icon" href="../../../gitbook/images/favicon.ico" type="image/x-icon">

    
    <link rel="next" href="Renderer.html" />
    
    

    </head>
    <body>
        
<div class="book">
    <div class="book-summary">
        
            
<div id="book-search-input" role="search">
    <input type="text" placeholder="Type to search" />
</div>

            
                <nav role="navigation">
                


<ul class="summary">
    
    

    

    
        
        
    
        <li class="chapter " data-level="1.1" data-path="../../../">
            
                <a href="../../../">
            
                    
                    Read Me
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2" data-path="../../Introduction.html">
            
                <a href="../../Introduction.html">
            
                    
                    1. Introduction
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.2.1" data-path="../../introduction/Motivation.html">
            
                <a href="../../introduction/Motivation.html">
            
                    
                    1.1. Motivation
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2.2" data-path="../../introduction/Principles.html">
            
                <a href="../../introduction/Principles.html">
            
                    
                    1.2. Principles
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2.3" data-path="../../introduction/Benefits.html">
            
                <a href="../../introduction/Benefits.html">
            
                    
                    1.3. Benefits
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2.4" data-path="../../introduction/Drawbacks.html">
            
                <a href="../../introduction/Drawbacks.html">
            
                    
                    1.4. Drawbacks
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2.5" data-path="../../introduction/Ecosystem.html">
            
                <a href="../../introduction/Ecosystem.html">
            
                    
                    1.5. Ecosystem
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2.6" data-path="../../introduction/Examples.html">
            
                <a href="../../introduction/Examples.html">
            
                    
                    1.6. Examples
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.3" data-path="../../Basics.html">
            
                <a href="../../Basics.html">
            
                    
                    2. Basics
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.3.1" data-path="../../basics/Rules.html">
            
                <a href="../../basics/Rules.html">
            
                    
                    2.1. Rules
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3.2" data-path="../../basics/Keyframes.html">
            
                <a href="../../basics/Keyframes.html">
            
                    
                    2.2. Keyframes
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3.3" data-path="../../basics/Fonts.html">
            
                <a href="../../basics/Fonts.html">
            
                    
                    2.3. Fonts
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3.4" data-path="../../basics/Renderer.html">
            
                <a href="../../basics/Renderer.html">
            
                    
                    2.4. Renderer
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.4" data-path="../../Advanced.html">
            
                <a href="../../Advanced.html">
            
                    
                    3. Advanced
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.4.1" data-path="../../advanced/DOMRendering.html">
            
                <a href="../../advanced/DOMRendering.html">
            
                    
                    3.1. DOM Rendering
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4.2" data-path="../../advanced/ServerRendering.html">
            
                <a href="../../advanced/ServerRendering.html">
            
                    
                    3.2. Server Rendering
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4.3" data-path="../../advanced/Plugins.html">
            
                <a href="../../advanced/Plugins.html">
            
                    
                    3.3. Plugins
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4.4" data-path="../../advanced/RendererConfiguration.html">
            
                <a href="../../advanced/RendererConfiguration.html">
            
                    
                    3.4. Renderer Configuration
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4.5" data-path="../../advanced/StaticStyle.html">
            
                <a href="../../advanced/StaticStyle.html">
            
                    
                    3.5. Global & Third-Party Styles
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4.6" data-path="../../advanced/CombinedRules.html">
            
                <a href="../../advanced/CombinedRules.html">
            
                    
                    3.6. Combined Rules
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4.7" data-path="../../advanced/Enhancers.html">
            
                <a href="../../advanced/Enhancers.html">
            
                    
                    3.7. Enhancers
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.5" data-path="../../UsageGuides.html">
            
                <a href="../../UsageGuides.html">
            
                    
                    4. Usage Guides
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.5.1" data-path="../../guides/UsageWithReact.html">
            
                <a href="../../guides/UsageWithReact.html">
            
                    
                    4.1. Usage with React
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.2" data-path="../../guides/UsageWithReactNative.html">
            
                <a href="../../guides/UsageWithReactNative.html">
            
                    
                    4.2. Usage with React Native
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.3" data-path="../../guides/UsageWithPreact.html">
            
                <a href="../../guides/UsageWithPreact.html">
            
                    
                    4.3. Usage with Preact
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.4" data-path="../../guides/UsageWithInferno.html">
            
                <a href="../../guides/UsageWithInferno.html">
            
                    
                    4.4. Usage with Inferno
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.5" data-path="../../guides/UsageWithAngular2.html">
            
                <a href="../../guides/UsageWithAngular2.html">
            
                    
                    4.5. Usage with Angular 2
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.6" data-path="../../Recipes.html">
            
                <a href="../../Recipes.html">
            
                    
                    5. Recipes
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.6.1" data-path="../../recipes/ResponsiveComponents.html">
            
                <a href="../../recipes/ResponsiveComponents.html">
            
                    
                    5.1. Responsive Components
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.2" data-path="../../recipes/TestingFelaComponents.html">
            
                <a href="../../recipes/TestingFelaComponents.html">
            
                    
                    5.2. Testing Fela Components
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.3" data-path="../../recipes/DisplayNameComponents.html">
            
                <a href="../../recipes/DisplayNameComponents.html">
            
                    
                    5.3. Explicit Component displayName
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.4" data-path="../../recipes/MonolithicClassNames.html">
            
                <a href="../../recipes/MonolithicClassNames.html">
            
                    
                    5.4. Monolithic Class Names
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.5" data-path="../../recipes/UsingRefWithReactComponent.html">
            
                <a href="../../recipes/UsingRefWithReactComponent.html">
            
                    
                    5.5. Using ref with React Component
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.7" data-path="../../API.html">
            
                <a href="../../API.html">
            
                    
                    6. API Reference
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.7.1" >
            
                <span>
            
                    
                    6.1. fela
            
                </span>
            

            
            <ul class="articles">
                
    
        <li class="chapter active" data-level="1.7.1.1" data-path="createRenderer.html">
            
                <a href="createRenderer.html">
            
                    
                    6.1.1. createRenderer
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.1.2" data-path="Renderer.html">
            
                <a href="Renderer.html">
            
                    
                    6.1.2. Renderer
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.1.3" data-path="combineRules.html">
            
                <a href="combineRules.html">
            
                    
                    6.1.3. combineRules
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.1.4" data-path="enhance.html">
            
                <a href="enhance.html">
            
                    
                    6.1.4. enhance
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.7.2" >
            
                <span>
            
                    
                    6.2. fela-dom
            
                </span>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.7.2.1" data-path="../fela-dom/render.html">
            
                <a href="../fela-dom/render.html">
            
                    
                    6.2.1. render
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.2.2" data-path="../fela-dom/renderToMarkup.html">
            
                <a href="../fela-dom/renderToMarkup.html">
            
                    
                    6.2.2. renderToMarkup
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.2.3" data-path="../fela-dom/renderToSheetList.html">
            
                <a href="../fela-dom/renderToSheetList.html">
            
                    
                    6.2.3. renderToSheetList
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.7.3" >
            
                <span>
            
                    
                    6.3. fela-native
            
                </span>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.7.3.1" data-path="../fela-native/createRenderer.html">
            
                <a href="../fela-native/createRenderer.html">
            
                    
                    6.3.1. createRenderer
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.3.2" data-path="../fela-native/Renderer.html">
            
                <a href="../fela-native/Renderer.html">
            
                    
                    6.3.2. Renderer
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.8" data-path="../../Troubleshooting.md">
            
                <span>
            
                    
                    7. Troubleshooting
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.9" data-path="../../FAQ.html">
            
                <a href="../../FAQ.html">
            
                    
                    8. FAQ
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.10" data-path="../../Feedback.html">
            
                <a href="../../Feedback.html">
            
                    
                    9. Feedback
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.11" data-path="../../Thanks.html">
            
                <a href="../../Thanks.html">
            
                    
                    10. Thanks
            
                </a>
            

            
        </li>
    

    

    <li class="divider"></li>

    <li>
        <a href="https://www.gitbook.com" target="blank" class="gitbook-link">
            Published with GitBook
        </a>
    </li>
</ul>


                </nav>
            
        
    </div>

    <div class="book-body">
        
            <div class="body-inner">
                
                    

<div class="book-header" role="navigation">
    

    <!-- Title -->
    <h1>
        <i class="fa fa-circle-o-notch fa-spin"></i>
        <a href="../../.." >6.1.1. createRenderer</a>
    </h1>
</div>




                    <div class="page-wrapper" tabindex="-1" role="main">
                        <div class="page-inner">
                            
<div id="book-search-results">
    <div class="search-noresults">
    
                                <section class="normal markdown-section">
                                
                                <h1 id="createrendererconfig"><code>createRenderer([config])</code></h1>
<p>Creates a Fela renderer which renders your selectors, keyframes, fonts and static styles.<br>
It caches all rendered styles to be able to reuse them on future rendering cycles.<br></p>
<h2 id="arguments">Arguments</h2>
<ol>
<li><code>config</code> (<em>Object?</em>): Optional renderer configuration. The most common use case adding <a href="../../advanced/Plugins.html">plugins</a> to process styles before they get cached. <em>(See <a href="../../advanced/RendererConfiguration.html">Advanced - Renderer Configuration</a> for further information)</em>.</li>
</ol>
<h2 id="returns">Returns</h2>
<p>(<a href="Renderer.html">Renderer</a>): A Renderer instance.</p>
<h2 id="example">Example</h2>
<pre class="language-"><code class="lang-javascript"><span class="token keyword">import</span> <span class="token punctuation">{</span> createRenderer <span class="token punctuation">}</span> <span class="token keyword">from</span> <span class="token string">&apos;fela&apos;</span>

<span class="token keyword">const</span> rule <span class="token operator">=</span> props <span class="token operator">=</span><span class="token operator">&gt;</span> <span class="token punctuation">(</span><span class="token punctuation">{</span>
  backgroundColor<span class="token punctuation">:</span> <span class="token string">&apos;red&apos;</span><span class="token punctuation">,</span>
  fontSize<span class="token punctuation">:</span> props<span class="token punctuation">.</span>size<span class="token punctuation">,</span>
  color<span class="token punctuation">:</span> <span class="token string">&apos;blue&apos;</span>
<span class="token punctuation">}</span><span class="token punctuation">)</span>

<span class="token keyword">const</span> renderer <span class="token operator">=</span> <span class="token function">createRenderer</span><span class="token punctuation">(</span><span class="token punctuation">)</span>

renderer<span class="token punctuation">.</span><span class="token function">render</span><span class="token punctuation">(</span>rule<span class="token punctuation">,</span> <span class="token punctuation">{</span> size<span class="token punctuation">:</span> <span class="token string">&apos;12px&apos;</span> <span class="token punctuation">}</span><span class="token punctuation">)</span> <span class="token comment" spellcheck="true">// =&gt; a b c</span>


console<span class="token punctuation">.</span><span class="token function">log</span><span class="token punctuation">(</span>renderer<span class="token punctuation">.</span><span class="token function">renderToString</span><span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
<span class="token comment" spellcheck="true">// .a{background-color:red}</span>
<span class="token comment" spellcheck="true">// .b{font-size:12px}</span>
<span class="token comment" spellcheck="true">// .c{color:blue}</span>
</code></pre>

                                
                                </section>
                            
    </div>
    <div class="search-results">
        <div class="has-results">
            
            <h1 class="search-results-title"><span class='search-results-count'></span> results matching "<span class='search-query'></span>"</h1>
            <ul class="search-results-list"></ul>
            
        </div>
        <div class="no-results">
            
            <h1 class="search-results-title">No results matching "<span class='search-query'></span>"</h1>
            
        </div>
    </div>
</div>

                        </div>
                    </div>
                
            </div>

            
                
                
                <a href="Renderer.html" class="navigation navigation-next navigation-unique" aria-label="Next page: 6.1.2. Renderer">
                    <i class="fa fa-angle-right"></i>
                </a>
                
            
        
    </div>

    <script>
        var gitbook = gitbook || [];
        gitbook.push(function() {
            gitbook.page.hasChanged({"page":{"title":"6.1.1. createRenderer","level":"1.7.1.1","depth":3,"next":{"title":"6.1.2. Renderer","level":"1.7.1.2","depth":3,"path":"docs/api/fela/Renderer.md","ref":"docs/api/fela/Renderer.md","articles":[]},"previous":{"title":"6.1. fela","level":"1.7.1","depth":2,"ref":"","articles":[{"title":"6.1.1. createRenderer","level":"1.7.1.1","depth":3,"path":"docs/api/fela/createRenderer.md","ref":"docs/api/fela/createRenderer.md","articles":[]},{"title":"6.1.2. Renderer","level":"1.7.1.2","depth":3,"path":"docs/api/fela/Renderer.md","ref":"docs/api/fela/Renderer.md","articles":[]},{"title":"6.1.3. combineRules","level":"1.7.1.3","depth":3,"path":"docs/api/fela/combineRules.md","ref":"docs/api/fela/combineRules.md","articles":[]},{"title":"6.1.4. enhance","level":"1.7.1.4","depth":3,"path":"docs/api/fela/enhance.md","ref":"docs/api/fela/enhance.md","articles":[]}]},"dir":"ltr"},"config":{"plugins":["edit-link","prism","-highlight","github","anker-enable","advanced-emoji"],"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"},"pluginsConfig":{"prism":{},"github":{"url":"https://github.com/rofrischmann/fela/"},"anker-enable":{},"search":{},"lunr":{"maxIndexSize":1000000},"fontsettings":{"theme":"white","family":"sans","size":2},"advanced-emoji":{"embedEmojis":false},"sharing":{"facebook":true,"twitter":true,"google":false,"weibo":false,"instapaper":false,"vk":false,"all":["facebook","google","twitter","weibo","instapaper"]},"edit-link":{"base":"https://github.com/rofrischmann/fela/tree/master","label":"Edit This Page"},"theme-default":{"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"},"showLevel":false}},"theme":"default","author":"Robin Frischmann <robin@rofrischmann.de>","pdf":{"pageNumbers":true,"fontSize":12,"fontFamily":"Arial","paperSize":"a4","chapterMark":"pagebreak","pageBreaksBefore":"/","margin":{"right":62,"left":62,"top":56,"bottom":56}},"structure":{"langs":"LANGS.md","readme":"README.md","glossary":"GLOSSARY.md","summary":"SUMMARY.md"},"variables":{},"title":"Fela","language":"en","gitbook":"3.1.0","description":"Official documentation for Fela"},"file":{"path":"docs/api/fela/createRenderer.md","mtime":"2017-09-19T09:16:24.000Z","type":"markdown"},"gitbook":{"version":"3.1.0","time":"2017-09-25T15:32:44.372Z"},"basePath":"../../..","book":{"language":""}});
        });
    </script>
</div>

        
    <script src="../../../gitbook/gitbook.js"></script>
    <script src="../../../gitbook/theme.js"></script>
    
        
        <script src="../../../gitbook/gitbook-plugin-edit-link/plugin.js"></script>
        
    
        
        <script src="../../../gitbook/gitbook-plugin-github/plugin.js"></script>
        
    
        
        <script src="../../../gitbook/gitbook-plugin-anker-enable/anker.js"></script>
        
    
        
        <script src="../../../gitbook/gitbook-plugin-search/search-engine.js"></script>
        
    
        
        <script src="../../../gitbook/gitbook-plugin-search/search.js"></script>
        
    
        
        <script src="../../../gitbook/gitbook-plugin-lunr/lunr.min.js"></script>
        
    
        
        <script src="../../../gitbook/gitbook-plugin-lunr/search-lunr.js"></script>
        
    
        
        <script src="../../../gitbook/gitbook-plugin-sharing/buttons.js"></script>
        
    
        
        <script src="../../../gitbook/gitbook-plugin-fontsettings/fontsettings.js"></script>
        
    

    </body>
</html>

