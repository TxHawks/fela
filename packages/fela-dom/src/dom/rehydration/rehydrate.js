/* @flow */
import arrayEach from 'fast-loops/lib/arrayEach'
import { RULE_TYPE } from 'fela-utils'

import rehydrateSupportRules from './rehydrateSupportRules'
import rehydrateRules from './rehydrateRules'

import type { DOMRenderer } from '../../../../../flowtypes/DOMRenderer'

// rehydration (WIP)
// TODO: static, keyframe, font
export default function rehydrate(renderer: DOMRenderer): void {
  arrayEach(document.querySelectorAll('[data-fela-type]'), node => {
    const rehydrationAttribute =
      node.getAttribute('data-fela-rehydration') || -1
    const rehydrationIndex = parseInt(rehydrationAttribute)

    // skip rehydration if no rehydration index is set
    // this index is set to -1 if something blocks rehydration
    if (rehydrationIndex !== -1) {
      const type = node.getAttribute('data-fela-type') || ''
      const media = node.getAttribute('media') || ''
      const support = node.getAttribute('data-fela-support')
      const css = node.textContent

      renderer.uniqueRuleIdentifier = rehydrationIndex

      if (type === RULE_TYPE) {
        if (support) {
          rehydrateSupportRules(css, media, renderer.cache)
        } else {
          rehydrateRules(css, media, '', renderer.cache)
        }
      }
    }
  })
}
