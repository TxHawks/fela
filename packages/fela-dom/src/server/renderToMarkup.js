/* @flow */
import {
  createStyleTagMarkup,
  objectReduce,
  clusterCache,
  RULE_TYPE,
  KEYFRAME_TYPE,
  FONT_TYPE,
  STATIC_TYPE
} from 'fela-utils'

const sheetMap = {
  fontFaces: FONT_TYPE,
  statics: STATIC_TYPE,
  keyframes: KEYFRAME_TYPE,
  rules: RULE_TYPE
}

export default function renderToMarkup(renderer: Object): string {
  const cacheCluster = clusterCache(renderer.cache, renderer.mediaQueryOrder)

  const basicMarkup = objectReduce(
    sheetMap,
    (markup, type, key) => {
      if (cacheCluster[key].length > 0) {
        markup += createStyleTagMarkup(cacheCluster[key], type)
      }

      return markup
    },
    ''
  )

  return objectReduce(
    cacheCluster.mediaRules,
    (markup, css, media) => {
      if (css.length > 0) {
        markup += createStyleTagMarkup(css, RULE_TYPE, media)
      }

      return markup
    },
    basicMarkup
  )
}
