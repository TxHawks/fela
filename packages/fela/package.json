{
  "name": "fela",
  "version": "6.0.0",
  "description": "Universal, Dynamic & High-Performance Styling in JavaScript",
  "typings": "index.d.ts",
  "main": "lib/index.js",
  "module": "es/index.js",
  "jsnext:main": "es/index.js",
  "scripts": {
    "postinstall": "opencollective postinstall"
  },
  "files": [
    "LICENSE",
    "README.md",
    "index.d.ts",
    "lib/**",
    "es/**"
  ],
  "repository": "https://github.com/rofrischmann/fela/",
  "keywords": [
    "fela",
    "dynamic styling",
    "stylesheet",
    "css",
    "styling",
    "cssinjs"
  ],
  "author": "Robin Frischmann",
  "license": "MIT",
  "dependencies": {
    "css-in-js-utils": "^2.0.0",
    "fela-utils": "^7.0.0",
    "opencollective": "^1.0.3"
  },
  "devDependencies": {
    "fela-tools": "^5.0.8"
  },
  "collective": {
    "type": "opencollective",
    "url": "https://opencollective.com/fela",
    "logo": "https://opencollective.com/opencollective/logo.txt"
  }
}
