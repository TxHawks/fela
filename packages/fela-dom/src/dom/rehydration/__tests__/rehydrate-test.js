import { createRenderer } from 'fela'
import webPreset from 'fela-preset-web'

import rehydrate from '../rehydrate'
import renderToMarkup from '../../../server/renderToMarkup'

const sortObject = obj =>
  Object.keys(obj)
    .sort()
    .reduce((newObj, key) => {
      newObj[key] = obj[key]
      return newObj
    }, {})

beforeEach(() => {
  const head = document.head
  while (head.firstChild) {
    head.removeChild(head.firstChild)
  }
})

describe('Rehydrating from DOM nodes', () => {
  it('should rehydrate the renderer cache', () => {
    const serverRenderer = createRenderer({
      filterClassName: cls => cls !== 'a',
      plugins: [...webPreset],
    })

    serverRenderer.renderRule(() => ({
      color: 'yellow',
      backgroundColor: 'red',
      flex: 1,
      '& #id > .foo ~ bar': {
        backgroundColor: 'red',
      },
      '[alt="Hello"]': {
        fontSize: 12,
      },
      '@supports (display: grid)': {
        color: 'blue',
        '&.foo.bar': {
          color: 'red',
        },
      },
      ':hover': {
        color: 'red',
        '> h1': {
          color: 'green',
        },
      },
    }))

    document.head.innerHTML = renderToMarkup(serverRenderer)

    const clientRenderer = createRenderer({
      filterClassName: cls => cls !== 'a',
      plugins: [...webPreset],
    })

    rehydrate(clientRenderer)

    expect([
      clientRenderer.uniqueRuleIdentifier,
      clientRenderer.cache,
    ]).toMatchSnapshot()

    expect(sortObject(clientRenderer.cache)).toEqual(
      sortObject(serverRenderer.cache)
    )
  })
})
