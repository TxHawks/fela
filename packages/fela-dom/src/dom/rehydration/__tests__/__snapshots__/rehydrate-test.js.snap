// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rehydrating from DOM nodes should rehydrate the renderer cache 1`] = `
Array [
  10,
  Object {
    " #id > .foo ~ barbackgroundColorred": Object {
      "className": "e",
      "declaration": "background-color:red",
      "media": "",
      "selector": ".e #id > .foo ~ bar",
      "support": "",
      "type": "RULE",
    },
    "(display: grid).foo.barcolorred": Object {
      "className": "h",
      "declaration": "color:red",
      "media": "",
      "selector": ".h.foo.bar",
      "support": "(display: grid)",
      "type": "RULE",
    },
    "(display: grid)colorblue": Object {
      "className": "g",
      "declaration": "color:blue",
      "media": "",
      "selector": ".g",
      "support": "(display: grid)",
      "type": "RULE",
    },
    ":hover> h1colorgreen": Object {
      "className": "j",
      "declaration": "color:green",
      "media": "",
      "selector": ".j:hover> h1",
      "support": "",
      "type": "RULE",
    },
    ":hovercolorred": Object {
      "className": "i",
      "declaration": "color:red",
      "media": "",
      "selector": ".i:hover",
      "support": "",
      "type": "RULE",
    },
    "WebkitFlex1;flex:1": Object {
      "className": "d",
      "declaration": "-webkit-flex:1;flex:1",
      "media": "",
      "selector": ".d",
      "support": "",
      "type": "RULE",
    },
    "[alt=\\"Hello\\"]fontSize12px": Object {
      "className": "f",
      "declaration": "font-size:12px",
      "media": "",
      "selector": ".f[alt=\\"Hello\\"]",
      "support": "",
      "type": "RULE",
    },
    "backgroundColorred": Object {
      "className": "c",
      "declaration": "background-color:red",
      "media": "",
      "selector": ".c",
      "support": "",
      "type": "RULE",
    },
    "coloryellow": Object {
      "className": "b",
      "declaration": "color:yellow",
      "media": "",
      "selector": ".b",
      "support": "",
      "type": "RULE",
    },
  },
]
`;
